dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NaiveLoopBoundAnnotation.java

```diff
@@ -0,0 +1,131 @@
+                    && liveAnalysis.getLiveRegistersAt(label).contains(init.getResultRegister())
+                    // Find the predicate and compute the bound
+                    && next instanceof Local ite && ite.getExpr() instanceof ITEExpr iteExpr
+                    && iteExpr.getCondition() instanceof IntCmpExpr cond
+                    && cond.getLeft().equals(init.getResultRegister())
+                    && (cond.getKind().equals(IntCmpOp.LT) || cond.getKind().equals(IntCmpOp.ULT)
+                            || cond.getKind().equals(IntCmpOp.LTE) || cond.getKind().equals(IntCmpOp.ULTE))
+                    && cond.getRight() instanceof IntLiteral bound
+                    // The exit condition is dependent on the counting register.
+                    && nNext instanceof CondJump continueJump && nnNext instanceof CondJump exitJump
+                    && exitJump.getLabel().equals(exit)
+                    && useDefAnalysis.getDefs(continueJump, ite.getResultRegister()).contains(ite)
+                    // There is a single increment to the register and that increment dominates the
+                    // loop backjump (this gives the step size).
+                    && getLoopBodyCountIncs(label, backJump, init.getResultRegister(), useDefAnalysis).count() == 1
+                    // The call to get() is guaranteed to succeed by the check above
+                    && preDominatorTree.isDominatedBy(backJump,
+                        getLoopBodyCountIncs(label, backJump, init.getResultRegister(), useDefAnalysis).findAny().get())
+            ) {
+                final Expression boundExpr = expressions.makeValue(
+                        // We use C-I+1 for i < C and C-I+2 for i <= C
+                        BigInteger.valueOf(bound.getValueAsInt() - initValExpr.getValueAsInt() + (cond.getKind().isStrict() ? 1 : 2)),
+                        bound.getType());
+                label.getPredecessor().insertAfter(EventFactory.Svcomp.newLoopBound(boundExpr));
+            }
+        }
+    }
+
+    private Stream<Event> getLoopBodyCountIncs(Label header, CondJump backJump, Register reg,
+            UseDefAnalysis useDefAnalysis) {
+        return header.getSuccessors().stream()
+                .filter(e -> e instanceof Local inc && inc.getExpr() instanceof IntBinaryExpr incExpr
+                        && incExpr.getLeft().equals(reg)
+                        && incExpr.getKind().equals(IntBinaryOp.ADD)
+                        && inc.getGlobalId() < backJump.getGlobalId()
+                        && e.getSuccessors().stream()
+                                .anyMatch(s -> s instanceof RegReader reader && s instanceof RegWriter writer
+                                        && s.getGlobalId() < backJump.getGlobalId()
+                                        && useDefAnalysis.getDefs(reader, inc.getResultRegister()).contains(inc)
+                                        && writer.getResultRegister().equals(reg)));
+    }
```

Comment:
These super long and nested statements are really not nice... and they tend to also be wrong :P. 
If there was any other operation on that register which is not an increment, you would not count it at all, although that other operation would likely break the whole argument

Wouldn't it be easier to first look for all `RegWriters` to `reg`. If there are multiple one's, you already lost (technically, if they are mutually exclusive you could have situations that are still fine but we ignore that).
Once you found the unique assignment, you traverse the Use-Def chain until you find an assignment that uses the same register (this is the cyclic data-flow). Along this chain, you probably want to only traverse it further back if the intermediary assignments are just copy-assignments (i.e., all these assignments do nothing) and there is only ever a unique Def (otherwise you get a Use-def-Tree rather than a chain). Alternatively, one could collect the effect along the Def-use chain to detect even more intricate cases.
Then you check what this unique assignment does: it must be an increment with some constant step size `S` (no need to restrict to `S==1`).

This is far more general and correct, and likely even easier to understand.
So in code like this (from your comment):
```
            DUMMY <- I
        Loop:
            r0 <- DUMMY // Useless, we can ignore
            r1 <- ITE((DUMMY < C), 1, 0)
            if (r1 != 0) then goto Continue
            goto Exit
        Continue:
            Body
            r2 <- (DUMMY + 1)
            DUMMY <- r2
            goto Loop
        Exit:
```
you would:
(1) Detect that `DUMMY` is interesting.
(2) Find the unique assignment `DUMMY <- r2`. Do a sanity check that this dominates the loop's backjump.
(3) Traverse the Use-Def chain and immediately see `r2 <- DUMMY + 1`. Since this uses only `DUMMY` and no other registers on the RHS, you check its shape and see that is of shape `+ 1`. Do a sanity check to see that his dominates the `DUMMY <- r2` assignment.
(4) Return the step-size as result of the computation (you can return `Optional<Integer>` with an empty optional if no step-size can be computed because, e.g., there are multiple assignments).
(5) The caller should have already found out the initial value of `DUMMY` and the Predicate (maybe also via separate functions?). Now he combines all this to compute a loop bound.

In a similar vein, you could detect the exit condition by chasing the Use-Def chain until you find an ITE of shape `ITE(reg < C, 0, 1)` to get your initial guess for the interesting register `reg`. This way, you would not rely on all these very specific code placement checks which would break as soon as we add annotations or instrumentation.

Reply:
> Wouldn't it be easier to first look for all `RegWriters` to `reg`. If there are multiple one's, you already lost (technically, if they are mutually exclusive you could have situations that are still fine but we ignore that). 

This I get.

> Once you found the unique assignment, you traverse the Use-Def chain until you find an assignment that uses the same register (this is the cyclic data-flow). 

Is there any way of doing this rather than:
- iterate over all `RegReader`s `reader` in the function
- check `useDefAnalysis.getDefs(reader, reg).contains(writer)`

> Along this chain, 

Which chain?

> you probably want to only traverse it further back if the intermediary assignments are just copy-assignments (i.e., all these assignments do nothing) and there is only ever a unique Def (otherwise you get a Use-def-Tree rather than a chain). 

I have no idea what this means.

> you would: (1) Detect that `DUMMY` is interesting. (2) Find the unique assignment `DUMMY <- r2`. Do a sanity check that this dominates the loop's backjump. (3) Traverse the Use-Def chain and immediately see `r2 <- DUMMY + 1`. 

Why wouldn't we get `r0 <- DUMMY` also here?

> In a similar vein, you could detect the exit condition by chasing the Use-Def chain until you find an ITE of shape `ITE(reg < C, 0, 1)` to get your initial guess for the interesting register `reg`. This way, you would not rely on all these very specific code placement checks which would break as soon as we add annotations or instrumentation.

You mean replacing lines ~91-100 which some method in a similar fashion as the above?

Reply:
ou found the unique assignment, you traverse the Use-Def chain until you find an assignment that uses the same register (this is the cyclic data-flow).
> 
> Is there any way of doing this rather than:
> 
>     * iterate over all `RegReader`s `reader` in the function
> 
>     * check `useDefAnalysis.getDefs(reader, reg).contains(writer)`

You have to find the unique assignment to the iteration count variable, which I assume you already have (e.g. by looking for the exit condition).
Once you have that assignment, you can check its RHS. If this RHS is just another register `i <- r` (`i` is the counter variable), you want to look for the assignment to `r` and you do this by using `useDef.getDefs(reader, r)` where `reader` is just `i <- r`. 
If this set is not a singleton, you have lost, so we assume you get a unique def.
This def can be yet another plain register assignment `r <- r2`, in which case you just repeat the procedure (this is the UseDef-chain you traverse!).
You do this until you find a non-trivial assignment and check its shape.
If its shape is `rk <- i + s` such that the loop variable is used, then you are done and you know that each iteration will set `i <- i + s` (btw. shouldn't SCCP's copy assignment collapse this chain already?).
If the assignment has any other shape (i.e., it's too complex), you just give up.
I hope this answers the next points as well.

> > you would: (1) Detect that `DUMMY` is interesting. (2) Find the unique assignment `DUMMY <- r2`. Do a sanity check that this dominates the loop's backjump. (3) Traverse the Use-Def chain and immediately see `r2 <- DUMMY + 1`.
>
> Why wouldn't we get `r0 <- DUMMY` also here?

You wouldn't because you `DUMMY <- r2` tells you to look for an assignment to `r2` (via UseDefAnalysis). The only one is `r2 <- DUMMY +1` and here you stop because you "completed the cycle".

> > In a similar vein, you could detect the exit condition by chasing the Use-Def chain until you find an ITE of shape `ITE(reg < C, 0, 1)` to get your initial guess for the interesting register `reg`. This way, you would not rely on all these very specific code placement checks which would break as soon as we add annotations or instrumentation.
> 
> You mean replacing lines ~91-100 which some method in a similar fashion as the above?

Yes. You would first look for all the jumps that exit the loop. There can be multiple one's, so one has to either check all of them or just naively pick one, most likely just the first one will do.
If you have the exiting jump, you check its condition and/or the condition of its predecessor (ideally one would collect the whole path condition from loop entry to exit, but whatever). If that exit condition depends on a singler register, you can start chasing its use-def chain until you find your `ITE(r < C, 0, 1)` of the desired shape.
This ITE gives you the predicate as well as the interesting register you need to check for.


Generally, all you need to manage is to find a candidate for the loop count variable and then check if its usage fits all desired properties: (1) initialized to some constant in the preheader, (2) incremented in each iteration (and nothing else), (3) used in the exit condition (you could have a function for each of those checks).
The best guess you get is by finding a register based on one of these properties and then check if it also satisfies the rest.
Currently your approach is centered around finding the exit condition and the registers it relies on, but you could also first look for the looping variables (the one's that are used before getting reassigned in the loop) instead.
It's up to you which approach you like better.